File: Fitness/ViewControllers/EditTrainingTableViewController.m
//
//  EditTrainingTableViewController.m
//  Fitness
//
//  Created by Mani on 23.04.15.
//  Copyright (c) 2015 Mani. All rights reserved.
//

#import "EditTrainingTableViewController.h"
#import "EditExerciseTableViewController.h"
#import "../Views/TrainingTableViewCell.h"
#import "../Resources/Training.h"
#import "../Resources/Exercise.h"

@interface EditTrainingTableViewController () <EditExerciseTableViewControllerDelegate, TrainingTableViewCellDelegate>

@end

@implementation EditTrainingTableViewController

#pragma mark Storyboard Actions
- (IBAction)onCancel:(id)sender {
    if(_delegate != nil)
        [_delegate editTrainingTableViewControllerDidCancelled:self];
}

- (IBAction)onSave:(id)sender {
    if(self.training.name.length == 0)
        return [[[UIAlertView alloc] initWithTitle:@"Fehler" message:@"Name ist leer!" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil] show];
    
    if(self.training.exercises.count == 0)
        return [[[UIAlertView alloc] initWithTitle:@"Fehler" message:@"Keine Übungen!" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil] show];
    
    if(_delegate != nil)
        [_delegate editTrainingTableViewControllerDidFinished:self];
}

#pragma mark Overloaded Base Methods
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.allowsSelectionDuringEditing = YES;
    self.tableView.editing = YES;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if([segue.destinationViewController isKindOfClass:[UINavigationController class]]) {
        UIViewController *underlayingController = ((UINavigationController *)segue.destinationViewController).topViewController;
        
        if([underlayingController isKindOfClass:[EditExerciseTableViewController class]]) {
            EditExerciseTableViewController *controller = (EditExerciseTableViewController *)underlayingController;
            controller.delegate = self;
            controller.sender = sender;
            
            if([sender isKindOfClass:[UITableViewCell class]]) {
                NSUInteger index = [self.tableView indexPathForCell:(UITableViewCell *)sender].row;
                
                Exercise *exercise = [self.training.exercises objectAtIndex:index];
                [controller setInitialData:exercise.type warmupTime:[exercise.warmup integerValue] timeInterval:[exercise.interval integerValue]];
            }
        }
    }
}

#pragma mark UITableView Data Source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 3;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    switch(section) {
        case 0:
            return 1;
        case 1:
            return self.training.exercises.count;
        case 2:
            return 1;
    }
    
    return 0;
}

#pragma mark UITableView Delegation
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSString *reuseIdentifier = @"";
    switch(indexPath.section) {
        case 0:
            reuseIdentifier = @"TrainingNameCell";
            break;
        case 1:
            reuseIdentifier = @"TrainingDetailCell";
            break;
        case 2:
            reuseIdentifier = @"AddTrainingCell";
            break;
    };
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:reuseIdentifier forIndexPath:indexPath];
   
    if(indexPath.section == 0) {
        TrainingTableViewCell *ptr = (TrainingTableViewCell *)cell;
        
        [ptr setInputText:self.training.name];
        ptr.delegate = self;
    }
    
    if(indexPath.section == 1) {
        Exercise *exercise = [self.training.exercises objectAtIndex:indexPath.row];
        
        NSInteger minutes = [exercise.interval integerValue] / 60;
        NSInteger seconds = [exercise.interval integerValue] % 60;
        
        cell.textLabel.text = [Exercise exerciseTypeToString:exercise.type];
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%ld Minute%@ %ld Sekunde%@", minutes, minutes == 1 ? @"" : @"n", seconds, seconds == 1 ? @"" : @"n"];
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if(indexPath.section != 1)
        return;
    
    UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
    [self performSegueWithIdentifier:@"EditExerciseTableViewController" sender:cell];
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return indexPath.section == 1; // Edit only exercises
}


- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle != UITableViewCellEditingStyleDelete)
        return;

    [self.training.exercises removeObjectAtIndex:indexPath.row];
    [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
}

- (NSIndexPath *)tableView:(UITableView *)tableView targetIndexPathForMoveFromRowAtIndexPath:(NSIndexPath *)sourceIndexPath toProposedIndexPath:(NSIndexPath *)proposedDestinationIndexPath
{
    if (sourceIndexPath.section != proposedDestinationIndexPath.section) {
        NSInteger row = 0;
        if (sourceIndexPath.section < proposedDestinationIndexPath.section) {
            row = [tableView numberOfRowsInSection:sourceIndexPath.section] - 1;
        }
        return [NSIndexPath indexPathForRow:row inSection:sourceIndexPath.section];
    }
    
    return proposedDestinationIndexPath;
}

- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
    id movee = [self.training.exercises objectAtIndex:fromIndexPath.row];
    [self.training.exercises removeObject:movee];
    [self.training.exercises insertObject:movee atIndex:toIndexPath.row];
}


- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    return indexPath.section == 1; // Move only exercises;
}

#pragma mark EditExerciseTableViewController Delegation
- (void)editExerciseTableViewController:(EditExerciseTableViewController *)controller didFinishedWithExerciseType:(enum ExerciseType)type withWarmupInterval:(NSTimeInterval)warmup withTimeInterval:(NSTimeInterval)interval {
    
    if(controller.sender == nil)
        return;

    // Edit the object only
    if([controller.sender isKindOfClass:[UITableViewCell class]]) {
        UITableViewCell *cell = (UITableViewCell *)controller.sender;
        NSIndexPath *path = [self.tableView indexPathForCell:cell];
        NSUInteger index = path.row;
        
        Exercise *exercise = [self.training.exercises objectAtIndex:index];
        exercise.type = type;
        exercise.warmup = [NSNumber numberWithInteger:warmup];
        exercise.interval = [NSNumber numberWithInteger:interval];
        
        // Disable the highlighting of the table's cell
        [self.tableView deselectRowAtIndexPath:path animated:YES];
    } else {
        Exercise *exercise = [[Exercise alloc] init];
        
        exercise.type = type;
        exercise.warmup = [NSNumber numberWithInteger:warmup];
        exercise.interval = [NSNumber numberWithInteger:interval];
        
        [self.training.exercises addObject:exercise];
    }
    
    [self.tableView reloadData];
    [controller dismissViewControllerAnimated:YES completion:nil];
    
}

- (void)editExerciseTableViewControllerDidCancel:(EditExerciseTableViewController *)controller {
    [controller dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark TrainingTableViewCell Delegation
- (void)trainingTableViewCell:(TrainingTableViewCell *)cell didChangedText:(NSString *)text {
    self.training.name = text;
}

@end

File: Fitness/ViewControllers/EditTrainingTableViewController.swift
//
//  EditTrainingTableViewController.swift
//  Fitness
//
//  Created by Manuel Stampfl on 24.06.15.
//  Copyright (c) 2015 mani1337. All rights reserved.
//

import UIKit

@objc protocol EditTrainingTableViewControllerDelegate {
    func editTrainingTableViewDidCancelled(controller: EditTrainingTableViewController)
    func editTrainingTableViewDidFinished(controller: EditTrainingTableViewController)
}

class EditTrainingTableViewController: UITableViewController, EditExerciseTableViewControllerDelegate, InputTableViewCellDelegate {
    // MARK: - Vars
    weak var delegate: EditTrainingTableViewControllerDelegate?
    var training = Training()
    var currentSelectedIndexPath: NSIndexPath?
    
    // Storyboard Actions
    @IBAction func onBarButtonItemSave(sender: AnyObject) {
        if self.training.name.isEmpty || self.training.exercises.isEmpty {
            let title = NSLocalizedString("TRAININGSTABLEVIEWCONTROLLER_INVALID_DATA_TITLE", comment: "title")
            let message = NSLocalizedString("TRAININGSTABLEVIEWCONTROLLER_INVALID_DATA_TEXT", comment: "text")
            let ack = NSLocalizedString("TRAININGSTABLEVIEWCONTROLLER_INVALID_DATA_ACK", comment: "ack")
            
            return UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: ack).show()
        }
        
        self.delegate?.editTrainingTableViewDidFinished(self)
    }
 
    @IBAction func onBarButtonItemClose(sender: AnyObject) {
        self.delegate?.editTrainingTableViewDidCancelled(self)
    }
    
    // MARK: - Overrided Base Methods
    override func viewDidLoad() {
        super.viewDidLoad()

        // Always in editing mode and allow cell selection while being in this mode
        self.editing = true
        self.tableView.allowsSelectionDuringEditing = true
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 3
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return 1
        }
        
        if section == 1 {
            return self.training.exercises.count
        }
        if section == 2 {
            return 1
        }
        
        return 0
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        if indexPath.section == 0 {
            let cell = tableView.dequeueReusableCellWithIdentifier("InputCell", forIndexPath: indexPath) as! InputTableViewCell
            cell.delegate = self
            cell.inputText = self.training.name
            return cell
        }
        
        if indexPath.section == 1 {
            let cell = tableView.dequeueReusableCellWithIdentifier("ExerciseCell", forIndexPath: indexPath) as! UITableViewCell
            
            var exercise = self.training.exercises[indexPath.row]
            let minutes = exercise.duration / 60
            let seconds = exercise.duration % 60
            
            cell.textLabel?.text = exercise.type.localizedName()
            cell.detailTextLabel?.text = String(format: NSLocalizedString("TRAININGSTABLEVIEWCONTROLLER_DURATION", comment: "duration"), arguments: [minutes, seconds])
            
            return cell
        }
        
        if indexPath.section == 2 {
            return tableView.dequeueReusableCellWithIdentifier("AddExerciseCell", forIndexPath: indexPath) as! UITableViewCell
        }
        
        // This point should never be reached so we don't care what we return
        return UITableViewCell()
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        self.currentSelectedIndexPath = indexPath
        
        if let vc: (UINavigationController, EditExerciseTableViewController) = self.storyboard?.instantiateNavigationControllerAndTopControllerWithIdentifier("EditExerciseTableViewController") {
            
            vc.1.delegate = self
            
            // Set the exercise
            if indexPath.section == 1 {
                vc.1.exercise = <-self.training.exercises[indexPath.row]
            }
            
            self.presentViewController(vc.0, animated: true, completion: nil)
        }
    }
  
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Only exercises can be edited
        return indexPath.section == 1
    }
 
    // Override to support editing the table view.
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if editingStyle != .Delete {
            return
        }
        
        self.training.exercises.removeAtIndex(indexPath.row)
        tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
    }

    override func tableView(tableView: UITableView, targetIndexPathForMoveFromRowAtIndexPath sourceIndexPath: NSIndexPath, toProposedIndexPath proposedDestinationIndexPath: NSIndexPath) -> NSIndexPath {
        if proposedDestinationIndexPath.section != 1 {
            return NSIndexPath(forRow: tableView.numberOfRowsInSection(sourceIndexPath.section) - 1, inSection: sourceIndexPath.section)
        }
        
        return proposedDestinationIndexPath
    }
    
    override func tableView(tableView: UITableView, moveRowAtIndexPath fromIndexPath: NSIndexPath, toIndexPath: NSIndexPath) {
        self.training.exercises.moveFromIndex(fromIndexPath.row, toIndex: toIndexPath.row)
    }

    override func tableView(tableView: UITableView, canMoveRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Only exercises can be moved
        return indexPath.section == 1
    }
    
    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case 0:
            return NSLocalizedString("TRAININGSTABLEVIEWCONTROLLER_SECTION_0", comment: "Section 0")
        case 1:
            return NSLocalizedString("TRAININGSTABLEVIEWCONTROLLER_SECTION_1", comment: "Section 1")
        default:
            return ""
        }
    }

    // MARK: - EditExerciseTableViewControllerDelegate
    func editExerciseTableViewControllerDidCancelled(controller: EditExerciseTableViewController) {
        controller.dismissViewControllerAnimated(true) {
            if self.currentSelectedIndexPath != nil {
                self.tableView.deselectRowAtIndexPath(self.currentSelectedIndexPath!, animated: true)
                self.currentSelectedIndexPath = nil
            }
        }
    }
    
    func editExerciseTableViewControllerDidFinished(controller: EditExerciseTableViewController) {
        if self.currentSelectedIndexPath != nil && controller.exercise != nil {
            if self.currentSelectedIndexPath!.section == 0 {
                self.training.exercises[self.currentSelectedIndexPath!.row] = controller.exercise!
            } else {
                self.training.exercises.append(controller.exercise!)
            }
        }
        
        controller.dismissViewControllerAnimated(true) {
            if self.currentSelectedIndexPath != nil {
                self.tableView.deselectRowAtIndexPath(self.currentSelectedIndexPath!, animated: true)
                self.currentSelectedIndexPath = nil
            }
        }
        
        tableView.reloadData()
    }
    
    // MARK: - InputTableViewCellDelegate
    func inputTableViewCell(cell: InputTableViewCell, didChangedText newText: String) {
        self.training.name = newText
    }
}
